cli_tools/common/utils/logging/logger.go

```diff
@@ -15,50 +15,179 @@
 package logging
 
 import (
-	"fmt"
 	"log"
+	"os"
+	"strings"
 	"time"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/proto/go/pb"
+	"github.com/golang/protobuf/proto"
 )
 
-// LogEntry encapsulates a single log entry.
-type LogEntry struct {
-	LocalTimestamp time.Time `json:"localTimestamp"`
-	Message        string    `json:"message"`
+// LogWriter is a logger for interactive tools. It supports
+// string messages and structured metrics.
+//
+// Structured metrics are accumulated over the lifespan of the logger.
+//
+// To rebuild the mock, run `go generate ./...`
+//go:generate go run github.com/golang/mock/mockgen -package mocks -source $GOFILE -destination ../../../mocks/mock_logger.go
+type LogWriter interface {
+	// WriteUser messages are shown in Pantheon or Gcloud to the user.
+	WriteUser(message string)
+	// WriteDebug messages are shown in the console output of the CLI tool.
+	WriteDebug(message string)
+	// WriteTrace messages are saved to the logging backend  (OutputInfo.serial_outputs)
+	WriteTrace(message string)
+	// WriteMetric merges all non-default fields into a single OutputInfo instance.
+	WriteMetric(metric *pb.OutputInfo)
+}
+
+// RedirectGlobalLogsToUser redirects the standard library's static logger
+// to logWriter.User.
+func RedirectGlobalLogsToUser(logWriter LogWriter) {
+	log.SetPrefix("")
+	log.SetFlags(0)
+	log.SetOutput(redirectShim{logWriter})
 }
 
-func newLogEntry(message string) *LogEntry {
-	return &LogEntry{LocalTimestamp: time.Now(), Message: message}
+// redirectShim forwards Go's standard logger to LogWriter.User.
+type redirectShim struct {
+	writer LogWriter
 }
 
-func (e *LogEntry) String() string {
-	return fmt.Sprintf("%s %s", e.LocalTimestamp.Format("2006-01-02T15:04:05Z"), e.Message)
+func (l redirectShim) Write(p []byte) (n int, err error) {
+	l.writer.WriteUser(string(p))
+	return len(p), nil
 }
 
-// LoggerInterface is logger abstraction
-type LoggerInterface interface {
-	Log(message string)
+// LogReader exposes pb.OutputInfo to a consumer.
+type LogReader interface {
+	ReadOutputInfo() *pb.OutputInfo
 }
 
-// NewStdoutLogger creates a new logger which uses prefix for all messages logged.
-// All messages are sent to stdout.
-func NewStdoutLogger(prefix string) LoggerInterface {
-	return stdoutLogger{prefix: prefix}
+// ToolLogger is a logger for interactive tools. It supports
+// string messages and structured metrics.
+//
+// Structured metrics are accumulated over the lifespan of the logger.
+type ToolLogger interface {
+	LogWriter
```

Comment:
I don't see an obvious benefits to split the interface to writer+reader. Would you explain the design? Just to hide the reader interface for CLI program?

Reply:
+1 on this

Reply:
Good question: It's a preventative measure to hide the information in `ReadOutputInfo`. Since this accumulates information, and will be available throughout the life of the request, there's a temptation for code to use these fields to perform control flow and logic. 

This creates hard-to-understand code, since "business information" can come from two places: function parameters, and this magical shared context.

Concretely, in other projects I've seen this manifest in as:
1. Putting  information that's required for logic into the global context, but shouldn't be logged. 
2. Dependencies are implicit, so re-ording tasks can break code that assumes a value will be set.
3. A developer thinks it's safe to modify metrics or logging, without knowing that some downstream code will be affected. Since the components are far away, this typically won't be found in unit tests. If it's found in integration tests, it typically results in hard-to-debug failures.

Reply:
The reading part is more like building the final result than just reading it, since it's a bit involved and it clears the caches. What about having the main `Logger` interface (that would do what `LogWriter` does) and `LogOutputInfoBuilder` that does what LogReader does.